File: apps/mobile_flutter/lib/core/network/host_validation.dart
/// Host validation for IPv4, hostname, and .local mDNS names.
library;

import 'dart:async';
import 'dart:io';

final _hostnamePattern = RegExp(
  r'^(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)(?:\.(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?))*$',
);
final _ipv4Pattern = RegExp(r'^\d{1,3}(?:\.\d{1,3}){3}$');

class HostValidationException implements Exception {
  HostValidationException(this.message);
  final String message;

  @override
  String toString() => message;
}

String normalizeServerHost(String raw) {
  final host = raw.trim();
  if (host.isEmpty) {
    throw HostValidationException('Host / IP address is required.');
  }
  if (_ipv4Pattern.hasMatch(host)) {
    return host;
  }
  final lowered = host.toLowerCase().replaceAll(RegExp(r'\.$'), '');
  if (!_hostnamePattern.hasMatch(lowered)) {
    throw HostValidationException(
      'Enter a valid IPv4 address, hostname, or mDNS name (for example WEBSTUDIO-SERVER.local).',
    );
  }
  return lowered;
}

String normalizeServerUrl(String raw, {int defaultPort = 8000}) {
  final trimmed = raw.trim();
  if (trimmed.isEmpty) {
    throw HostValidationException('Server address is required.');
  }

  var candidate = trimmed;
  if (!candidate.startsWith('http://') && !candidate.startsWith('https://')) {
    candidate = 'http://$candidate';
  }

  final uri = Uri.parse(candidate);
  final host = normalizeServerHost(uri.host);
  final port = uri.hasPort ? uri.port : defaultPort;
  return Uri(scheme: 'http', host: host, port: port).toString().replaceAll(RegExp(r'/+$'), '');
}

String? extractHostnameFromUrl(String url) {
  try {
    return normalizeServerHost(Uri.parse(url).host);
  } catch (_) {
    return null;
  }
}

Future<HostResolutionResult> resolveServerHost(String host, {int port = 8000}) async {
  try {
    final normalized = normalizeServerHost(host);
    if (_ipv4Pattern.hasMatch(normalized)) {
      return HostResolutionResult(
        configuredHost: normalized,
        resolvedIp: normalized,
        success: true,
        message: 'Using IPv4 address $normalized.',
      );
    }

    final results = await InternetAddress.lookup(
      normalized,
      type: InternetAddressType.IPv4,
    ).timeout(
      const Duration(seconds: 5),
      onTimeout: () => throw TimeoutException('Timed out resolving $normalized.'),
    );
    if (results.isEmpty) {
      return HostResolutionResult(
        configuredHost: normalized,
        resolvedIp: null,
        success: false,
        message: 'Could not resolve $normalized.',
      );
    }
    final ip = results.first.address;
    return HostResolutionResult(
      configuredHost: normalized,
      resolvedIp: ip,
      success: true,
      message: 'Resolved $normalized to $ip.',
    );
  } on TimeoutException catch (error) {
    return HostResolutionResult(
      configuredHost: host.trim(),
      resolvedIp: null,
      success: false,
      message: error.message ?? 'Timed out resolving host.',
    );
  } on HostValidationException catch (error) {
    return HostResolutionResult(
      configuredHost: host.trim(),
      resolvedIp: null,
      success: false,
      message: error.message,
    );
  } on SocketException catch (error) {
    return HostResolutionResult(
      configuredHost: host.trim(),
      resolvedIp: null,
      success: false,
      message: error.message,
    );
  }
}

class HostResolutionResult {
  const HostResolutionResult({
    required this.configuredHost,
    required this.resolvedIp,
    required this.success,
    required this.message,
  });

  final String configuredHost;
  final String? resolvedIp;
  final bool success;
  final String message;
}

File: apps/backend/src/webstudio_backend/core/network/host_validation.py
"""Server host validation — IPv4, hostname, and .local mDNS names."""

from __future__ import annotations

import asyncio
import ipaddress
import re
import socket
from dataclasses import dataclass

_HOSTNAME_PATTERN = re.compile(
    r"^(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)(?:\.(?:[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?))*$",
)


class HostValidationError(ValueError):
    pass


@dataclass(slots=True)
class HostResolution:
    configured_host: str
    resolved_ip: str | None
    success: bool
    message: str


def normalize_server_host(raw: str) -> str:
    host = raw.strip()
    if not host:
        raise HostValidationError("Host / IP address is required.")
    if len(host) > 253:
        raise HostValidationError("Host / IP address is too long.")
    try:
        ipaddress.ip_address(host)
        return host
    except ValueError:
        pass
    lowered = host.lower().removesuffix(".")
    if not _HOSTNAME_PATTERN.fullmatch(lowered):
        raise HostValidationError(
            "Enter a valid IPv4 address, hostname, or mDNS name (for example WEBSTUDIO-SERVER.local).",
        )
    return lowered


async def resolve_server_host(host: str, *, port: int = 8000) -> HostResolution:
    configured = normalize_server_host(host)
    try:
        loop = asyncio.get_running_loop()
        results = await loop.getaddrinfo(
            configured,
            port,
            type=socket.SOCK_STREAM,
            family=socket.AF_INET,
        )
        if not results:
            return HostResolution(
                configured_host=configured,
                resolved_ip=None,
                success=False,
                message=f"Could not resolve {configured}.",
            )
        resolved_ip = results[0][4][0]
        return HostResolution(
            configured_host=configured,
            resolved_ip=resolved_ip,
            success=True,
            message=f"Resolved {configured} to {resolved_ip}.",
        )
    except socket.gaierror as exc:
        return HostResolution(
            configured_host=configured,
            resolved_ip=None,
            success=False,
            message=f"DNS lookup failed for {configured}: {exc.strerror or exc}.",
        )
    except HostValidationError as exc:
        return HostResolution(
            configured_host=host.strip(),
            resolved_ip=None,
            success=False,
            message=str(exc),
        )
